tools/stage_time_analysis.groovy

```diff
@@ -0,0 +1,82 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import groovy.json.JsonSlurper
+
+  def getJobs(String stats_file, String find_stage, String folder, Object list_jobs) {
+    list_jobs.each { job ->
+      if (job.jobs && job.jobs.size() > 0) {
+        if (folder.equals("")) {
+          getJobs(stats_file, find_stage, "job/${job.name}", job.jobs)
+        } else {
+          getJobs(stats_file, find_stage, "${folder}/job/${job.name}", job.jobs)
+        }
+      } else {
+        def jobName="${folder}/job/${job.name}"
+        if (folder.equals("")) {
+          jobName="job/${job.name}"
+        }
+        println("${jobName}")
+
+        def workflow = null
+        try {
+          workflow = sh(returnStdout: true, script: "wget -q -O - ${params.JENKINS_URL}/${jobName}/lastSuccessfulBuild/wfapi/describe")
+        } catch(Exception e) {
+          workflow = null
+        }
+        if (workflow) {
+          def json = new JsonSlurper().parseText(workflow)
+          json.stages.each { stage ->
+            if (stage.name.equalsIgnoreCase(find_stage)) {
+              def duration = stage.durationMillis/60000
+              duration = duration.intValue()
+              println("  ==> Job: ${jobName} Stage: ${stage.name} durationMinutes: ${duration}")
+              sh("echo \'${jobName}:${duration}\' >> ${stats_file}")
+            }
+          }
+        } else {
+          println("  ==> no lastSuccessfulBuild")
+        }
+      }
+    }
+  }
+
+node ("master") {
+  def jenkinsUrl = "${params.JENKINS_URL}"
+  def find_stage = "${params.STAGE}"
+  def stats_file = "stage.stats"
+
+  stage("getStageStats") {
+    sh("rm -f ${stats_file}")
+
+    try {
+      // Get all jobs list
+      def jobs = sh(returnStdout: true, script: "wget -q -O - ${jenkinsUrl}/api/json?tree=jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name,jobs[name]]]]]]]]]]")
```

Comment:
This seems brittle (the chain of name,job, name, job...

Reply:
Yeah, the API is somewhat limited to return sub-folders, i've allowed a depth of 10 folders, which is probably reasonable.

Reply:
Cool, can you enhance the comment explain thy we can't have nice things? :-)